File: exercise7/Exercise 7 - From images to localization/code/parabolaRansac.m
function [best_guess_history, max_num_inliers_history] = ...
    parabolaRansac(data, max_noise)
% data is 2xN with the data points given column-wise, 
% best_guess_history is 3xnum_iterations with the polynome coefficients 
%   from polyfit of the BEST GUESS SO FAR at each iteration columnwise and
% max_num_inliers_history is 1xnum_iterations, with the inlier count of the
%   BEST GUESS SO FAR at each iteration.
rng(2);
N = size(data,2);
iter_num = 100;
max_num_inliers = 0;
max_num_inliers_history = zeros(1,iter_num);
best_guess_history = zeros(3,iter_num);

for i = 1:iter_num
    sample_pt = datasample(data',3,'Replace',false); % 3*2
% X = [sample_pt(1,1)^2, sample_pt(1,1), 1;
%      sample_pt(2,1)^2, sample_pt(2,1), 1;
%      sample_pt(3,1)^2, sample_pt(3,1), 1];
% Y = [sample_pt(1,2);sample_pt(2,2);sample_pt(3,2)];
% para = (X'*X)\A'*Y;
    x = sample_pt(:,1);
    y = sample_pt(:,2);
    p = polyfit(x,y,2);

% compute error
    x_all = data(1,:);
    y_all = data(2,:);

    y_predict = polyval(p,x_all);
    y_delta = abs(y_predict - y_all);

    k = find(y_delta < max_noise);
    num_inliers = size(k,2);
    if num_inliers > max_num_inliers
        max_num_inliers = num_inliers;
        max_num_inliers_history(i) = max_num_inliers;
        best_guess_history(:,i) = p;
    else
        max_num_inliers_history(i) = max_num_inliers_history(i-1);
        best_guess_history(:,i) = best_guess_history(:,i-1);
    end
end


end

File: exercise7_solution/Solution 7 - From images to localization/code_python/parabolaRansac.py
import numpy as np

def parabolaRansac(data, max_noise):
    """
    best_guess_history is 3xnum_iterations with the polynome coefficients  
    from polyfit of the BEST GUESS SO FAR at each iteration columnwise and max_num_inliers_history is
    1xnum_iterations, with the inlier count of the BEST GUESS SO FAR at each iteration.
    """
    num_iterations = 100

    best_guess_history = np.zeros([3, num_iterations])
    max_num_inliers_history = np.zeros([num_iterations])

    best_guess = np.zeros([3, 1])
    max_num_inliers = 0
    rerun_on_inliers = True

    for i in range(num_iterations):
        # Model based on 3 samples:
        indices = np.random.permutation(data.shape[1])[:3]
        samples = data[:, indices]
        guess = np.polyfit(samples[0, :], samples[1,:], 2)

        # Evaluate amount of inliers
        errors = np.abs(np.polyval(guess, data[0, :]) - data[1,:])
        inliers = errors <= max_noise + 1e-5
        num_inliers = (inliers).sum()
        # Determine if the current guess is the best so far.
        if num_inliers > max_num_inliers:
            if rerun_on_inliers:
                guess = np.polyfit(data[0, inliers], data[1, inliers], 2)
            best_guess = guess
            max_num_inliers = num_inliers

        best_guess_history[:, i] = best_guess
        max_num_inliers_history[i] = max_num_inliers

    return best_guess_history, max_num_inliers_history
